src/linear_algebra/alternating.lean

```diff
@@ -0,0 +1,222 @@
+import group_theory.perm.sign
+
+/-!
+# Alternating Maps
+
+We construct the bundled function `alternating_map`, which extends `multilinear_map` with all the
+arguments of the same type.
+
+## Main definitions
+* `alternating_map R M N ι` is the space of `R`-linear alternating maps from `ι → M` to `N`.
+* `f.map_eq_args` expresses that `f` is zero when two inputs are equal.
+* `f.map_swap` expresses that `f` is negated when two inputs are swapped.
+* `f.map_perm` expresses how `f` varies by a sign change under a permutation of its inputs.
+* An `add_comm_monoid`, `add_comm_group`, and `semimodule` structure over `alternating_map`s that
+  matches the definitions over `multilinear_map`s.
+
+## Implementation notes
+`alternating_map` is defined in terms of `map_eq_args`, as this is easier to work with than
+using `map_swap` as a definition, and does not require `has_neg N`.
+-/
+
+-- semiring / add_comm_monoid
+variables {R : Type*} [semiring R]
+variables {M : Type*} [add_comm_monoid M] [semimodule R M]
+variables {N : Type*} [add_comm_monoid N] [semimodule R N]
+
+-- semiring / add_comm_group
+variables {L : Type*} [add_comm_group L] [semimodule R L]
+
+-- ring / add_comm_group
+variables {R' : Type*} [ring R']
+variables {M' : Type*} [add_comm_group M'] [semimodule R' M']
+variables {N' : Type*} [add_comm_group N'] [semimodule R' N']
+
+variables {ι : Type*} [decidable_eq ι]
+
+set_option old_structure_cmd true
+
+section
+variables (R M N ι)
+/--
+An alternating map is a multilinear map that vanishes when two of its arguments are equal.
+-/
+structure alternating_map extends multilinear_map R (λ i : ι, M) N :=
+(map_eq_args' : ∀ (v : ι → M) (i j : ι) (h : v i = v j) (hij : i ≠ j), to_fun v = 0)
+end
+
+attribute [nolint doc_blame] alternating_map.to_multilinear_map
+
+namespace alternating_map
+
+variables (f f' : alternating_map R M N ι)
+variables (g' : alternating_map R' M' N' ι)
+variables (v : ι → M) (v' : ι → M')
+open function
+
+/-! Basic coercion simp lemmas, largely copied from `ring_hom` -/
+section coercions
+
+instance : has_coe_to_fun (alternating_map R M N ι) := ⟨_, λ x, x.to_fun⟩
+
+initialize_simps_projections alternating_map (to_fun → apply)
+
+@[simp] lemma to_fun_eq_coe : f.to_fun = f := rfl
+
+@[simp] lemma coe_mk (f : (ι → M) → N) (h₁ h₂ h₃) : ⇑(⟨f, h₁, h₂, h₃⟩ :
+  alternating_map R M N ι) = f := rfl
+
+instance : has_coe (alternating_map R M N ι) (multilinear_map R (λ i : ι, M) N) :=
+⟨λ x, x.to_multilinear_map⟩
+
+@[simp, norm_cast] lemma coe_multilinear_map : ⇑(f : multilinear_map R (λ i : ι, M) N) = f := rfl
+
+@[simp] lemma to_multilinear_map_eq_coe : f.to_multilinear_map = f := rfl
+
+@[simp] lemma coe_multilinear_map_mk (f : (ι → M) → N) (h₁ h₂ h₃) :
+  ((⟨f, h₁, h₂, h₃⟩ : alternating_map R M N ι) :  multilinear_map R (λ i : ι, M) N) = ⟨f, h₁, h₂⟩ :=
+rfl
+
+@[ext] theorem ext {f f' : alternating_map R M N ι} (H : ∀ x, f x = f' x) : f = f' :=
+by cases f; cases f'; congr'; exact funext H
```

Comment:
```suggestion
by { cases f, cases f', congr', exact funext H }
```
I thought this was the preferred style now.

Reply:
This was copied verbatim from an (outdated) copy of multilinear map. I suppose there's no real argument not to update it though.